### ISSUE TITLE
jax.disable_jit() magically resolves NaNs

### ISSUE THREAD

**Issue by samuela** [google/jax#3443]
I have a bit of code that runs producing a NaN gradient in the very first iteration. I use `@jit` extensively for performance. Running the code normally produces a NaN error: 
```
  File "/root/research/research/odecontrol/lqr_colloc_fwd_step_bwd.py", line 96, in bwd
    assert tree_all(tree_map(lambda x: jnp.all(jnp.isfinite(x)), aug))
AssertionError
```
but if I just wrap my entire main function in a `with jax.disable_jit():`, the code runs without any NaNs whatsoever.

Is this behavior up to spec? What could possibly be causing the discrepancy?

**Comment by shoyer** [comment]
XLA can fuse together and rewrite operations under a JIT, which can sometimes lead to different numerics. It’s likely that that is the source of your observed discrepancy.

**Comment by samuela** [comment]
@shoyer I see, that's probably what's going on here then. Floating point arithmetic is annoying...

**Comment by mattjj** [comment]
+1 to what @shoyer said! IIUC that answers the question as to whether `jit` can affect float numerics.

By the way, do you know about [jax_debug_nans](https://github.com/google/jax/pull/482#issue-258999834)? (#2649 is an issue to document it better.) It might help narrow down where the nan is occurring, but given that the nans disappear when jit is disabled, it might not.

Let us know if you think we can help figure out your nan issue. Since this issue thread is (AIUI) narrowly about whether disable_jit can change nan behavior, and that question is answered, I'm going to close it.
